File: js/domain/cartridge/SegaMasterSystemCartridge.js
/* 
 * Project: EGGStation - Sega Master System Emulator
 * Author: Enrique González Gutiérrez
 * 
 * Domain Layer: Sega Master System Cartridge Entity
 * 
 * Represents a physical Master System Cartridge. Handles raw binary loading, 
 * copier header removal, metadata validation, and checksum generation.
 */

class SegaMasterSystemCartridge {
    /**
     * @param {string} filename - The filename of the ROM image.
     */
    constructor(filename) {
        this.cartridgeSize = 0;
        this.cartridgeRom = [];
        this.cartridgeName = filename;
        this.romChecksum = 0;
    }

    /**
     * Checks for the presence of the standard SEGA security registration string ("TMR SEGA").
     * @param {string[]} headerCharacters - The 8 characters retrieved from address 0x7FF0.
     * @returns {boolean} True if standard Sega header is present.
     */
    checkForTmrSega(headerCharacters) {
        const tmrSegaSignature = ['T', 'M', 'R', ' ', 'S', 'E', 'G', 'A'];

        for (let i = 0; i < 8; i++) {
            if (headerCharacters[i] !== tmrSegaSignature[i]) {
                return false;
            }
        }

        return true;
    }

    /**
     * Diagnostically prints regional hardware specifications based on ROM metadata.
     * @param {number} regionCode - 4-bit regional code.
     */
    printRegionCode(regionCode) {
        if (regionCode === 0x03) {
            console.log("Cartridge::Region Code decoded: SMS Japan");
        } else if (regionCode === 0x04) {
            console.log("Cartridge::Region Code decoded: SMS Export");
        }
    }

    /**
     * Diagnostically prints the ROM storage size based on internal metadata.
     * @param {number} sizeCode - 4-bit size classification code.
     */
    printRomSize(sizeCode) {
        if (sizeCode === 0x0c) {
            console.log("Cartridge::Internal ROM size value: 32KB");
        } else if (sizeCode === 0x0f) {
            console.log("Cartridge::Internal ROM size value: 128KB");
        } else if (sizeCode === 0x0) {
            console.log("Cartridge::Internal ROM size value: 256KB");
        }
    }

    /**
     * Calculates the standard 32-bit checksum of the internal ROM byte array.
     * @returns {number} Unsigned 32-bit checksum value.
     */
    calculateChecksum() {
        let checksum = 0;

        if ((this.cartridgeRom.length % 4) !== 0) {
            throw new Error("Cartridge::Error: ROM length is not a multiple of 4, cannot calculate checksum.");
        }

        for (let i = 0; i < this.cartridgeRom.length; i += 4) {
            let chunk = this.cartridgeRom[i];
            chunk |= this.cartridgeRom[i + 1] << 8;
            chunk |= this.cartridgeRom[i + 2] << 16;
            chunk |= this.cartridgeRom[i + 3] << 24;

            checksum += chunk;
            checksum &= 0xffffffff; // Force unsigned 32-bit math wrap-around      
        }

        return Math.abs(checksum);
    }

    /**
     * Parses the raw ArrayBuffer, extracts metadata, and prepares ROM arrays.
     * @param {ArrayBuffer} buffer - Raw file array buffer.
     */
    load(buffer) {
        this.cartridgeSize = buffer.byteLength;
        const tempArray = new Uint8Array(buffer);

        for (let i = 0; i < this.cartridgeSize; i++) {
            this.cartridgeRom.push(tempArray[i]);
        }

        if (this.cartridgeSize < (32 * 1024)) {
            console.warn("Cartridge::Warning: Loaded ROM size is smaller than the standard 32KB system minimum.");
        }

        // Clean outdated copier prefixes (e.g. game doctor headers) if detected
        if (this.cartridgeRom.length % 0x4000 === 512) {
            console.log("Cartridge::Detected and removed 512-byte copier header.");
            const cleanedRom = [];
            for (let i = 512; i < this.cartridgeRom.length; i++) {
                cleanedRom.push(this.cartridgeRom[i]);
            }
            this.cartridgeRom = cleanedRom;
        }

        // Compute unique game checksum identification
        this.romChecksum = this.calculateChecksum();
        console.log(`Cartridge::Checksum identity calculated as: 0x${this.romChecksum.toString(16).padStart(8, '0')}`);

        // Validate Standard Sega registration block at address 0x7FF0
        const headerBytes = [];
        for (let i = 0; i < 16; i++) {
            headerBytes.push(this.cartridgeRom[0x7ff0 + i]);
        }

        const characters = [];
        headerBytes.forEach(byte => {
            characters.push(String.fromCharCode(byte));
        });

        if (this.checkForTmrSega(characters)) {
            console.log("Cartridge::Standard SEGA registration signature confirmed at address 0x7FF0.");
        } else {
            console.warn("Cartridge::No standard SEGA signature found. Proceeding with flat fallback mode.");
            return;
        }

        this.printRegionCode(headerBytes[0x0f] >> 4);
        this.printRomSize(headerBytes[0x0f] & 0x0f);
    }
}

File: src/domain/SegaMasterSystemCartridge.cpp
/**
 * Project: EGGStation - Sega & SNES Multi-System Emulator
 * Author: Enrique González Gutiérrez
 * File: src/domain/SegaMasterSystemCartridge.cpp
 * 
 * Domain Layer: Sega Master System Cartridge Entity
 * 
 * Role:
 * Implementation of the SegaMasterSystemCartridge entity.
 * Handles the binary parsing, checksum calculation, and copier header stripping 
 * logic for physical Sega Master System ROM files.
 */

#include "SegaMasterSystemCartridge.h"
#include <iostream>

SegaMasterSystemCartridge::SegaMasterSystemCartridge() 
    : cartridgeSize(0), romChecksum(0), cartridgeName("") {}

void SegaMasterSystemCartridge::load(const std::string& filename, const uint8_t* rawData, uint32_t size) {
    uint32_t startOffset = 0;

    // Detect and strip standard 512-byte copier headers (e.g., Game Doctor backup headers)
    if (size % 0x4000 == 512) {
        startOffset = 512;
        std::cout << "[Cartridge] Detected and stripped 512-byte copier header." << std::endl;
    }

    cartridgeRom.clear();
    uint32_t cleanSize = size - startOffset;
    cartridgeRom.resize(cleanSize);
    
    // Copy the binary payload safely into the internal domain vector
    memcpy(cartridgeRom.data(), rawData + startOffset, cleanSize);
    cartridgeSize = cleanSize;
    cartridgeName = filename;

    // Generate unique ROM checksum identity
    romChecksum = calculateChecksum();
    std::cout << "[Cartridge] Checksum calculated successfully: 0x" 
              << std::hex << romChecksum << std::dec << std::endl;

    // Validate standard Sega hardware registration header at 0x7FF0
    if (checkForTmrSegaHeader()) {
        std::cout << "[Cartridge] Verified SEGA physical registration header at offset 0x7FF0." << std::endl;
    } else {
        std::cout << "[Cartridge] Warning: No standard SEGA registration header detected." << std::endl;
    }
}

bool SegaMasterSystemCartridge::checkForTmrSegaHeader() const {
    if (cartridgeRom.size() < 0x8000) {
        return false; // ROM is too small to contain a standard header
    }

    const char tmrSegaSignature[8] = {'T', 'M', 'R', ' ', 'S', 'E', 'G', 'A'};
    
    for (int i = 0; i < 8; i++) {
        if (cartridgeRom[0x7FF0 + i] != static_cast<uint8_t>(tmrSegaSignature[i])) {
            return false;
        }
    }
    return true;
}

uint32_t SegaMasterSystemCartridge::calculateChecksum() {
    uint32_t checksum = 0;
    uint32_t length = cartridgeRom.size();

    // Enforce 32-bit dword-aligned loops. 
    // This replicates the unsigned 32-bit accumulation of the JavaScript implementation.
    for (uint32_t i = 0; i < length; i += 4) {
        uint32_t chunk = cartridgeRom[i];
        if (i + 1 < length) chunk |= (cartridgeRom[i + 1] << 8);
        if (i + 2 < length) chunk |= (cartridgeRom[i + 2] << 16);
        if (i + 3 < length) chunk |= (cartridgeRom[i + 3] << 24);

        checksum += chunk; // Naturally wraps around on 32-bit overflow
    }
    return checksum;
}

// --- Getters ---

const std::vector<uint8_t>& SegaMasterSystemCartridge::getRomData() const {
    return cartridgeRom;
}

uint32_t SegaMasterSystemCartridge::getCartridgeSize() const {
    return cartridgeSize;
}

uint32_t SegaMasterSystemCartridge::getRomChecksum() const {
    return romChecksum;
}

const std::string& SegaMasterSystemCartridge::getCartridgeName() const {
    return cartridgeName;
}